code/network/multi_interpolate.h

```diff
@@ -0,0 +1,127 @@
+	int _local_skip_forward;			// TODO: Use this? Used on the server if a client is inexplicably ahead of us.
+	bool _simulation_mode;				// are we currently simulating as if this were a normal ship?
+	bool _packets_expended;				// have we used up all packets while interpolating?
+	SCP_vector<packet_info> _packets;	// all the info from the position/orientation portion of packets that we care to keep
+	int _source_player_index;
+
+	void reassess_packet_index(vec3d* pos, matrix* ori, physics_info* pip);		// for finding which packets from within _packets we should use
+	void replace_packet(int index, vec3d* pos, matrix* orient, physics_info* pip);	// a function that acts as a workaround, when coming out of simulation_mode
+
+	// Frame numbers that helps us figure out if we should ignore new information coming from the server because
+	// we already received a newer packet than this one for that type of info.
+	int _hull_comparison_frame;						// what frame was the last hull information received?
+	int _shields_comparison_frame;					// what frame was the last shield information received?
+	SCP_vector<int> _subsystems_comparison_frame;	// what frame was the last subsystem information received? (for each subsystem)
+	int _ai_comparison_frame;						// what frame was the last ai information received?
+
+public:
+
+	// adds a new packet, whilst also manually sorting the relevant entries
+	void add_packet(int frame, int time_delta, vec3d* position, vec3d* velocity, vec3d* rotational_velocity, vec3d* desired_velocity, vec3d* desired_rotational_velocity, angles* angles, int player_index);
+	void interpolate(vec3d* pos, matrix* ori, physics_info* pip, bool player_ship);
+
+	int get_hull_comparison_frame() { return _hull_comparison_frame; }
+	int get_shields_comparison_frame() { return _shields_comparison_frame; }
+	
+	int get_subsystem_comparison_frame(int i) 
+	{ 
+		if (i < static_cast<int>(_subsystems_comparison_frame.size()) && i >= 0) {
+			return _subsystems_comparison_frame[i]; 
+		} // if it somehow got passed nonsense, chances are what it is is trying to read is nonsense, and INT_MAX will keep it from doing anything crazy.
+		else {
+			return INT_MAX;
+		}
+	}
+
+	int get_ai_comparison_frame() { return _ai_comparison_frame; }
+
+	void set_hull_comparison_frame(int frame) { _hull_comparison_frame = frame; }
+	void set_shields_comparison_frame(int frame) { _shields_comparison_frame = frame; }
+
+	void set_subsystem_comparison_frames(int i, int frame)
+	{ 
+		if (i < static_cast<int>(_subsystems_comparison_frame.size()) && i >= 0) {
+			_subsystems_comparison_frame[i] = frame;
+		} 
+	}
+
+	void set_ai_comparison_frame(int frame) { _ai_comparison_frame = frame; }
+
+	void force_interpolation_mode() { _simulation_mode = true; }
+
+	void reset() 
+	{
+		_packets.clear();
+		_packets.reserve(PACKET_INFO_LIMIT * 2);
+		_upcoming_packet_index = -1;
```

Comment:
There should be a multi mode check here so that it will bail if single player. That way single player mode doesn't waste time/memory dealing with this interpolation stuff.